Dont decompose aten.baddmm in inductor

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137904

Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D65218743](https://our.internmc.facebook.com/intern/diff/D65218743)

eellison committed:
Dont decompose aten.baddmm in inductor

[ghstack-poisoned]

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ngimel commented on test/inductor/test_max_autotune.py:
```diff
@@ -578,6 +578,32 @@ def f(x, weight):
+                return torch.ops.aten.baddbmm.default(self.bias, x, self.weight)
+
+        x = torch.randn(
+            64, 2048, 64, dtype=torch.float16, requires_grad=False, device="cuda"
```
I'd be happier if test had specially constructed inputs such that bmm would e.g. overflow but `beta<1` would again bring it to the correct range - that way it'd be a clear failure if we were again to mess up decomposition, rather than "the accuracy is worse than original op"

eellison replied:
Surprisingly, even beta, alpha == 1 was sufficient to cause numerics issue in the @SherlockNoMad 's not upcasted decomposed pr:

```
Exception: Caused by sample input at index 0: SampleInput(input=Tensor[size=(5, 5, 10), device="cuda:0", dtype=torch.bfloat16], args=TensorList[Tensor[size=(5, 5, 5), device="cuda:0", dtype=torch.bfloat16], Tensor[size=(5, 5, 10), device="cuda:0", dtype=torch.bfloat16]], kwargs={'beta': '1', 'alpha': '1'}, broadcasts_input=False, name='')
```

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "failing internal tests D64418200, some results not within tolerance?" -c ghfirst

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "unfortunately the failures on the previous import are still present on the current one D64568703" -c ghfirst

clee2000 commented:
@pytorchbot revert -m "unfortunately the failures on the previous import are still present on the current one D64568703" -c ghfirst

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65075947).

eellison commented:
@pytorchbot merge -i

huydhn commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but I think the test is failing in trunk, maybe a landrace?' -c nosignal

inductor/test_torchinductor.py::CpuTests::test_baddbmm_cpu [GH job link](https://github.com/pytorch/pytorch/actions/runs/11565238726/job/32195498097) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/c066f4a055020ae994dd10a1b1fafbe3774108cd)

desertfire commented:
> @pytorchbot revert -m 'Sorry for reverting your change but I think the test is failing in trunk, maybe a landrace?' -c nosignal
> 
> inductor/test_torchinductor.py::CpuTests::test_baddbmm_cpu [GH job link](https://github.com/pytorch/pytorch/actions/runs/11565238726/job/32195498097) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/c066f4a055020ae994dd10a1b1fafbe3774108cd)

@eellison , I see it failed on a cpp wrapper test added by me. It can be resolved by adding `aten.baddbmm.out` to https://github.com/pytorch/pytorch/blob/main/torchgen/aoti/fallback_ops.py and rerun torchgen.

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65218743).

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65218743).

eellison commented:
@pytorchbot rebase

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
cc @deertfire, does the fallback calling not work with kwargs ?

eellison commented:
@pytorchbot rebase

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65218743](https://our.internmc.facebook.com/intern/diff/D65218743)

[ghstack-poisoned]

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65218743](https://our.internmc.facebook.com/intern/diff/D65218743)

[ghstack-poisoned]

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -4577,6 +4577,10 @@ def __init__(  # type: ignore[no-untyped-def]
         self.set_cpp_kernel_name(cpp_kernel_name)
         self.set_python_kernel_name(python_kernel_name)
+        if not ordered_kwargs_for_cpp_kernel and op_overload is not None:
+            ordered_kwargs_for_cpp_kernel = tuple(
```
`collect_arg_kwarg_properties` already sets `ordered_kwargs_for_cpp_kernel`. The new CI failures seem relate to  that.

eellison committed:
Update on "Dont decompose aten.baddmm in inductor"


Previously the decomposition would upcasts inputs to fp32. This led to a slowdown compared to eager which would run in fp16. We also tried keeping the bmm in fp16, and the upcasting for the epilogue but that led to worse numerics because the bmm in eager would do the epilogue all in fp32 without a downcast in the bmm accumulator.

Fix for https://github.com/pytorch/pytorch/issues/137897


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65218743](https://our.internmc.facebook.com/intern/diff/D65218743)

[ghstack-poisoned]

eellison commented:
Apparently i cant reland this since internal diff was closed and i cant unlink. New pr: https://github.com/pytorch/pytorch/pull/141045
